src/assembler.cpp

```diff
@@ -1508,4 +1521,83 @@ void Assembler::ResolveLabelOffsets(Label* label) {
     }
 }
 
+template <typename T>
+void Assembler::PlaceAtOffset(Literal<T>* literal, Literal<T>::LocationOffset offset) {
+    BISCUIT_ASSERT(literal != nullptr);
+    BISCUIT_ASSERT(offset >= 0 && offset <= m_buffer.GetCursorOffset());
+
+    const T& value = literal->Place(offset);
+    ResolveLiteralOffsets(literal);
+    literal->ClearOffsets();
+
+    m_buffer.Emit(value);
+}
+
+template <typename T>
+ptrdiff_t Assembler::LinkAndGetOffset(Literal<T>* literal) {
+    BISCUIT_ASSERT(literal != nullptr);
+
+    // If we have a placed literal, then it's straightforward to calculate
+    // the offsets.
+    if (literal->IsPlaced()) {
+        const auto cursor_address = m_buffer.GetCursorAddress();
+        const auto literal_offset = m_buffer.GetOffsetAddress(*literal->GetLocation());
+        return static_cast<ptrdiff_t>(literal_offset - cursor_address);
+    }
+
+    // If we don't have a placed literal, we return an offset of zero.
+    // While the emitter will emit a bogus load instruction initially,
+    // the offset will be patched over once the literal has been properly
+    // placed at a location.
+    literal->AddOffset(m_buffer.GetCursorOffset());
+    return 0;
+}
+
+template <typename T>
+void Assembler::ResolveLiteralOffsets(Literal<T>* literal) {
+    const auto is_auipc_type = [](uint32_t instruction) {
+        return (instruction & 0x7F) == 0b0010111;
+    };
+
+    const auto is_gpr_load_type = [](uint32_t instruction) {
+        return (instruction & 0x7F) == 0b0000011;
+    };
+
+    const auto literal_location = *literal->GetLocation();
+
+    for (const auto offset : literal->m_offsets) {
+        const auto address = m_buffer.GetOffsetAddress(offset);
+        auto* const ptr = reinterpret_cast<uint8_t*>(address);
+
+        uint32_t instructions[2] = {0, 0};
```

Comment:
generally prefer std::array when working outside of C interfaces (and honestly with them, for the most part):

```suggestion
        std::array<uint32_t, 2> instructions{};
```